rclpy/src/rclpy/context.cpp

```diff
@@ -149,13 +152,20 @@ Context::shutdown()
 {
   {
     std::lock_guard guard{g_contexts_mutex};
-    g_contexts.erase(
-      std::remove(g_contexts.begin(), g_contexts.end(), rcl_context_.get()),
-      g_contexts.end());
+    if (std::find(g_contexts.begin(), g_contexts.end(), rcl_context_.get()) != g_contexts.end()) {
+      g_contexts.erase(
+        std::remove(g_contexts.begin(), g_contexts.end(), rcl_context_.get()),
+        g_contexts.end());
+    }
   }
-  rcl_ret_t ret = rcl_shutdown(rcl_context_.get());
-  if (RCL_RET_OK != ret) {
-    throw RCLError("failed to shutdown");
+  // Just in case, check if context is valid
+  if (rcl_context_is_valid(rcl_context_.get())) {
+    rcl_ret_t ret = rcl_shutdown(rcl_context_.get());
+    if (RCL_RET_OK != ret) {
+      throw RCLError("failed to shutdown");
+    }
+  } else {
+    throw RCLError("context is already shutdown");
```

Comment:
I would recommend against `RCLError` since `rcl_context_is_valid` won't set the error state, meaning the message will be something like error not set. See how `RCLError` reads the error message here: https://github.com/ros2/rclpy/blob/31147813b653b4820c0b7c16b12ee30f8ab29ccb/rclpy/src/rclpy/exceptions.cpp#L56

But also, see the `rcl_shutdown` comment above